packages/web-components/fast-components/src/design-system-provider/README.md

```diff
@@ -0,0 +1,251 @@
+
+Notice below that none of the code-samples initialize a value to a declared _design system_ property. This is because the _design system_ is designed to function as a waterfall of values - where values assigned lower in the DOM tree override values defined by ancestors. Initializing the value on the instance assigns the value to every instance of the _Design System Provider_ which is generally un-desired.
+
+**EXAMPLE: initialized vs non-initialized properties**
+
+```ts
+class ExampleClass {
+    public nonInitialized;
+    // vs
+    public initialized = true;
+}
+```
+
+The `use-defaults` boolean attribute exposes a mechanism to apply the default values to an element while still allowing nested design system elements to intentionally override specific values. For details on how to set default values, see [Declaring Design System Properties](#Declaring-Design-System-Properties)
+
+### Composing Design System Providers
+*Design System Providers* are designed to compose their values with ancestor *Design System Providers* to facilitate changing values for decedents of the provider.
+
+```html
+<fast-design-system-provider design-unit="4" background-color="#FFFFFF">
+    <p style="height: calc(var(--design-unit) * 1px)">4px</p>
+    <p style="background: var(--background-color)">#FFFFFF</p>
+
+    <fast-design-system-provider design-unit="8">
+        <p style="height: calc(var(--design-unit) * 1px)">8px</p>
+        <p style="background: var(--background-color)">#FFFFFF</p>
+    </fast-design-system-provider>
+</fast-design-system-provider>
+```
+
+
+### Recipes
+
+There are some things that CSS just can't do. Advanced math, array manipulation, and conditionals are simply not possible with today's CSS.
+
+To address this gap, the _DesignSystemProvider_ is capable of registering `CSSCustomPropertyDefinition` types through the `registerCSSCustomProperty` method:
+
+**EXAMPLE: Register a CSS custom property that is a function of the _design system_**
+
+```ts
+DesignSystemProvider.registerCSSCustomProperty({
+    name: "design-unit-12th",
+    value: designSystem => Math.pow(designSystem.designUnit, 12),
+});
+```
+
+In the above example, the value function will be re-evaluated if the _design system_ ever changes.
+
+The above API is made especially useful when defining a [component stylesheet](https://github.com/microsoft/fast-dna/blob/master/packages/web-components/fast-element/docs/building-components.md#defining-css) - individual stylesheets can declare dependencies on CSS custom properties that are functions of the element instance's *design system*
+
+**EXAMPLE: creating a recipe dependency**
+```ts
+import { css } from "@microsoft/fast-element";
+import { cssCustomPropertyBehaviorFactory, DesignSystemProvider } from "@microsoft/fast-components";
```

Comment:
I'm not able to get `cssCustomPropertyBehaviorFactory` to pull in from the package (I'm testing locally by using `yarn link` in conjunction with #3051):

```
TS2305: Module '"../../../../../../packages/web-components/fast-components/dist"' has no exported member 'cssCustomPropertyBehaviorFactory'.
```

Reply:
Hmmm it might not be exported actually - good find.

Reply:
This should be resolved now by #3085